frame/message-queue/src/lib.rs

```diff
@@ -929,6 +946,85 @@ impl<T: Config> Pallet<T> {
 		Ok(())
 	}
 
+	pub fn do_discard_overweight(
+		origin: MessageOriginOf<T>,
+		page_index: PageIndex,
+		message_index: T::Size,
+	) -> Result<(), Error<T>> {
+		let (page, book_state, message_pos, message) =
+			Self::extract_overweight(&origin, page_index, message_index)?;
+		// A post-dispatch refund is not needed since we don't process the message.
+		let _page_removed = Self::remove_overweight(
+			origin.clone(),
+			book_state,
+			page,
+			page_index,
+			message_pos,
+			message.len(),
+		)?;
+		let hash = T::Hashing::hash(&message);
+		Self::deposit_event(Event::OverweightDiscarded { hash, origin, page_index, message_index });
+
+		Ok(())
+	}
+
+	/// Extract an overweight message and return `(Page, Book, position, message)`.
+	fn extract_overweight(
+		queue: &MessageOriginOf<T>,
+		page_index: PageIndex,
+		message_index: T::Size,
+	) -> Result<(PageOf<T>, BookStateOf<T>, usize, Vec<u8>), Error<T>> {
+		let book_state = BookStateFor::<T>::get(queue);
+		let page = Pages::<T>::get(queue, page_index).ok_or(Error::<T>::NoPage)?;
+		let (pos, is_processed, payload) =
+			page.peek_index(message_index.into() as usize).ok_or(Error::<T>::NoMessage)?;
+
+		ensure!(!is_processed, Error::<T>::AlreadyProcessed);
+		ensure!(
+			page_index < book_state.begin ||
+				(page_index == book_state.begin && pos < page.first.into() as usize),
+			Error::<T>::Queued
+		);
+
+		let msg = payload.to_vec();
+		Ok((page, book_state, pos, msg))
+	}
+
+	/// Remove an overweight message.
+	///
+	/// Returns whether the page was removed.
+	fn remove_overweight(
```

Comment:
the function removes also just processed message, not necessary overweight message

Reply:
Good spot!